### ISSUE TITLE
TypeError: Cannot read property '_requestId' of undefined from chromePerflogParser.js

### ISSUE THREAD

**Issue by dylangrafmyre** [sitespeedio/browsertime#209]
I am unable to attempt Chrome Version 54.0.2840.87, 3 runs without crashing with ` TypeError: Cannot read property '_requestId' of undefined`. I consistently crash on the first run. 

Just testing against `www.friendsandguests.com` without the custom login works fine, but using the custom login and watching the specific url render properly in the browser seems all is well until it fails within the `chromePerflogParser` module.

Any ideas what I am doing wrong? 

Thanks for the great set of tools!

```
➜  dylan@DYLANs-MacBook-Pro-3:~/sitespeed  sitespeed.io --preScript login.js -vv -b chrome  https://www.friendsandguests.com/s                                     1 ↵ [20:33:24]
[2016-11-04 20:37:02] Versions OS: darwin 14.5.0 nodejs: v6.2.2 sitespeed.io: 4.0.3 browsertime: 1.0.0-beta.9 coach: 0.28.4
[2016-11-04 20:37:02] Starting chrome for analysing https://www.friendsandguests.com/s 3 time(s)
[2016-11-04 20:37:02] Testing url https://www.friendsandguests.com/s run 1
[2016-11-04 20:37:11] Requesting url https://www.friendsandguests.com/s
[2016-11-04 20:37:14] Waiting for script 'return (function() {try { var end = window.performance.timing.loadEventEnd;return (end > 0) && (Date.now() > end + 2000);} catch(e) {return true;}})()' at most 300000 ms
[2016-11-04 20:37:16] Executing script userAgent
[2016-11-04 20:37:16] Executing script windowSize
[2016-11-04 20:37:16] Executing script documentHeight
[2016-11-04 20:37:16] Executing script documentTitle
[2016-11-04 20:37:16] Executing script documentWidth
[2016-11-04 20:37:16] Executing script responsive
[2016-11-04 20:37:16] Executing script firstPaint
[2016-11-04 20:37:16] Executing script fullyLoaded
[2016-11-04 20:37:16] Executing script navigationTiming
[2016-11-04 20:37:16] Executing script pageTimings
[2016-11-04 20:37:16] Executing script resourceTimings
[2016-11-04 20:37:16] Executing script rumSpeedIndex
[2016-11-04 20:37:16] Executing script userTimings
[2016-11-04 20:37:16] Executing script coachAdvice
[2016-11-04 20:37:17] Getting performance logs from Chrome
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Network loading finished for requestId 61979.201 with no matching request.
[2016-11-04 20:37:17] Recieved network response for requestId 61979.200 with no matching request.
[2016-11-04 20:37:17] Recieved network data for requestId 61979.200 with no matching request.
[2016-11-04 20:37:17] Network loading finished for requestId 61979.200 with no matching request.
[2016-11-04 20:37:17] Failed to load url: https://maps.gstatic.com/mapfiles/transparent.png
[2016-11-04 20:37:17] Executing script CHROME_NAME_AND_VERSION_JS
[2016-11-04 20:37:17] Executing script PAGE_TITLE_JS
[2016-11-04 20:37:17] Telling browser to quit.
[2016-11-04 20:37:17] Testing url https://www.friendsandguests.com/s run 2
[2016-11-04 20:37:25] Requesting url https://www.friendsandguests.com/s
[2016-11-04 20:37:28] Waiting for script 'return (function() {try { var end = window.performance.timing.loadEventEnd;return (end > 0) && (Date.now() > end + 2000);} catch(e) {return true;}})()' at most 300000 ms
[2016-11-04 20:37:30] Executing script userAgent
[2016-11-04 20:37:30] Executing script windowSize
[2016-11-04 20:37:30] Executing script documentHeight
[2016-11-04 20:37:30] Executing script documentTitle
[2016-11-04 20:37:30] Executing script documentWidth
[2016-11-04 20:37:30] Executing script responsive
[2016-11-04 20:37:30] Executing script firstPaint
[2016-11-04 20:37:30] Executing script fullyLoaded
[2016-11-04 20:37:30] Executing script navigationTiming
[2016-11-04 20:37:30] Executing script pageTimings
[2016-11-04 20:37:30] Executing script resourceTimings
[2016-11-04 20:37:30] Executing script rumSpeedIndex
[2016-11-04 20:37:30] Executing script userTimings
[2016-11-04 20:37:30] Executing script coachAdvice
[2016-11-04 20:37:31] Getting performance logs from Chrome
[2016-11-04 20:37:31] Telling browser to quit.
[2016-11-04 20:37:31] TypeError: Cannot read property '_requestId' of undefined
    at Object.harFromEvents (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/support/chromePerflogParser.js:235:26)
    at runner.getLogs.timeout.map.tap.then (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engineDelegate/chromeDelegate.js:70:39)
From previous event:
    at ChromeDelegate.onStopIteration (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engineDelegate/chromeDelegate.js:70:8)
    at runner.start.tap.tap.tap.then.then.tap.tap (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engine.js:204:39)
From previous event:
    at runIteration (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engine.js:204:10)
    at Promise.reduce (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engine.js:242:25)
From previous event:
    at Promise.resolve.tap.tap.tap (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engine.js:241:17)
From previous event:
    at Engine.run (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/browsertime/lib/core/engine.js:240:8)
    at engine.start.then (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/lib/plugins/browsertime/analyzer.js:76:26)
From previous event:
    at Object.analyzeUrl (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/lib/plugins/browsertime/analyzer.js:76:8)
    at Object.processMessage (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/lib/plugins/browsertime/index.js:129:27)
    at queue.process (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/lib/support/queueHandler.js:157:57)
    at drainItem (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/concurrent-queue/index.js:92:21)
    at Immediate.drain [as _onImmediate] (/Users/dylan/.nvm/versions/node/v6.2.2/lib/node_modules/sitespeed.io/node_modules/concurrent-queue/index.js:73:84)
    at tryOnImmediate (timers.js:543:15)
    at processImmediate [as _immediateCallback] (timers.js:523:5)
[2016-11-04 20:37:31] Render HTML for 1 page(s)
[2016-11-04 20:37:32] HTML stored in /Users/dylan/sitespeed/sitespeed-result/www.friendsandguests.com/2016-11-04-20-37-02
[2016-11-04 20:37:32] Finished analysing https://www.friendsandguests.com/s
```

**Comment by dylangrafmyre** [comment]
I cloned `browsertime` and did some local testing with some log statements where it was failing at 
`lib/support/chromePerflogParser.js:235:26`.

It is failing for me at `let previousEntry = entries.find((entry) => entry._requestId === params.requestId);` because as the logs point out there is only `1 entry` in the `Entries array` and there is no match for `entry._requestId === params.requestId`. 

`previousEntry` is undefined due to no match from within the `Entries array`.

What is the best way to move on if the `.find` method call comes up empty? 

Thanks,
Dylan

```
➜  dylan@DYLANs-MacBook-Pro-3:~/browsertime (master) ✗ bin/browsertime.js --preScript login.js https://www.friendsandguests.com/s                                                      1 ↵ ✹ ✭ [14:18:20]
[2016-11-05 14:22:06] Running chrome for url: https://www.friendsandguests.com/s
[2016-11-05 14:22:06] Testing url https://www.friendsandguests.com/s run 1
[2016-11-05 14:22:19] Current Entry requestId:  72880.208
[2016-11-05 14:22:19] Params requestId:  72880.208
[2016-11-05 14:22:19] Entries Length:  1
[2016-11-05 14:22:19] Entry requestId from Entries Array:  72880.210
[2016-11-05 14:22:20] Error running browsertime TypeError: Cannot read property '_requestId' of undefined
    at Object.harFromEvents (/Users/dylan/browsertime/lib/support/chromePerflogParser.js:244:26)
    at runner.getLogs.timeout.map.tap.then (/Users/dylan/browsertime/lib/core/engineDelegate/chromeDelegate.js:70:39)
```

**Comment by beenanner** [comment]
@dylangrafmyre does it work without the preScript. Everything seems to work fine for me on that URL without it.

```
➜  browsertime git:(master) ✗ node -v
v6.2.2
➜  browsertime git:(master) ✗ npm -v
3.9.5

➜  browsertime git:(master) ✗ bin/browsertime.js -b chrome https://www.friendsandguests.com/s 
[2016-11-05 23:33:51] Running chrome for url: https://www.friendsandguests.com/s
[2016-11-05 23:33:51] Testing url https://www.friendsandguests.com/s run 1
[2016-11-05 23:33:56] Testing url https://www.friendsandguests.com/s run 2
[2016-11-05 23:34:00] Testing url https://www.friendsandguests.com/s run 3
[2016-11-05 23:34:05] 28 requests, 1409.03 kb, firstPaint: 1.54s (±0.04), DOMContentLoaded: 1.54s (±0.04), Load: 1.86s (±0.04), rumSpeedIndex: 1.63s (±0.04) (3 runs)
[2016-11-05 23:34:05] Wrote data to browsertime-results/www.friendsandguests.com-s/2016-11-05T233351-0400
```

**Comment by dylangrafmyre** [comment]
@beenanner Thanks for testing! 
You cannot actually access `https://www.friendsandguests.com/s` without logging into the site.

Most likely your test is succeeding due to testing the redirect page `https://www.friendsandguests.com/users/sign_in` which you will get redirected to when non-authenticated users try to access `https://www.friendsandguests.com/s`.

As you can see from my first comment, I was able to successfully complete 1 fun before failing. Which to me points as the preScript not being a issue?. If the preScript was an issue, shouldn't it always fail on the first fun?

Also, when you authenticate with the preScript and test the url `https://www.friendsandguests.com/s` with have some third party js for our Intercom service that is running for the Intercom Messenger that renders on the page. This is what is catching the if statement `if (params.redirectResponse)` within `chromePerflogParser.js` and failing when no matching `entry_requestID === params.requestID` from the `entries.find` method call.

**Comment by beenanner** [comment]
Thanks for the explanation @dylangrafmyre do you mind sharing a gist with your preScript ( minus the login username/password  ;-p ) to help reproduce and debug?

**Comment by dylangrafmyre** [comment]
@beenanner If i add conditional assignment to `chomePerflogParser.js:235`, I am able to successfully complete 3 runs with the preScript authentication process. 
```
if (params.redirectResponse) {
            let previousEntry = entries.find((entry) => entry._requestId === params.requestId);
            previousEntry = (typeof previousEntry === "undefined") ? entry : previousEntry
            previousEntry._requestId += 'r';
            populateEntryFromResponse(previousEntry, params.redirectResponse);
          }
```

Feel free to create an account on `www.friendsandguests.com` if you want to test the complete process.

Here is the login script: https://gist.github.com/dylangrafmyre/9d32b051ae2d571f32fba7561eeac320

**Comment by beenanner** [comment]
Cool! Thanks for the gist. First off I want thank you @dylangrafmyre we really appreciate you taking time to help work through this issue. Secondly feel free to submit a pull-request with that patch and either @tobli, @soulgalore or myself should be able to review and test.

**Comment by beenanner** [comment]
@dylangrafmyre testing your script + patch I have confirmed it completes all 3 runs now where without it fails on the second run as you posted above, but I do see with the patch some missed requests that I need to dig into. 

```
➜  browsertime git:(master) ✗ bin/browsertime.js --preScript preScript.js -b chrome https://www.friendsandguests.com/s
[2016-11-06 00:39:29] Running chrome for url: https://www.friendsandguests.com/s
[2016-11-06 00:39:29] Testing url https://www.friendsandguests.com/s run 1
[2016-11-06 00:39:40] Testing url https://www.friendsandguests.com/s run 2
[2016-11-06 00:39:51] Recieved network response for requestId 1719.211 with no matching request.
[2016-11-06 00:39:51] Recieved network data for requestId 1719.211 with no matching request.
[2016-11-06 00:39:51] Network loading finished for requestId 1719.211 with no matching request.
[2016-11-06 00:39:51] Recieved network response for requestId 1719.212 with no matching request.
[2016-11-06 00:39:51] Recieved network data for requestId 1719.212 with no matching request.
[2016-11-06 00:39:51] Network loading finished for requestId 1719.212 with no matching request.
[2016-11-06 00:39:51] Testing url https://www.friendsandguests.com/s run 3
[2016-11-06 00:40:02] Recieved network response for requestId 1878.211 with no matching request.
[2016-11-06 00:40:02] Recieved network data for requestId 1878.211 with no matching request.
[2016-11-06 00:40:02] Network loading finished for requestId 1878.211 with no matching request.
[2016-11-06 00:40:02] Recieved network response for requestId 1878.212 with no matching request.
[2016-11-06 00:40:02] Recieved network data for requestId 1878.212 with no matching request.
[2016-11-06 00:40:02] Network loading finished for requestId 1878.212 with no matching request.
[2016-11-06 00:40:03] 18 requests, 33.17 kb, firstPaint: 2.18s (±0.09), DOMContentLoaded: 1.92s (±0.09), Load: 2.s (±0.), rumSpeedIndex: 2.49s (±0.) (3 runs)
[2016-11-06 00:40:03] Wrote data to browsertime-results/www.friendsandguests.com-s/2016-11-06T003929-0400

```

**Comment by beenanner** [comment]
Ok @tobli more background info on this. It appears that there are a few issues going on here after my testing. Looks like it is only capturing the network requests from the first pages in this case the login page instead of the page after the preScript.js logs in the user in the actual HAR file. I don't think that issue is specific to this ticket so going to file an additional ticket to investigate how best we want to handle this use-case.

**Comment by dylangrafmyre** [comment]
@beenanner Thank you so much for all the debugging and investigating this issue!! Again thank you for the great tools.

**Comment by beenanner** [comment]
Absolutely @dylangrafmyre :-)

**Comment by tobli** [comment]
Sorry for the late reply. This is how I understand this (might not be the complete picture):
- the current design is to only include the actual tested page in the har, not the pre/post scripts. This is so tools analysing the actual page (such as sitespeed.io) can easily know what to look at in the har (i.e. don't need to somehow filter out pre/post actions).
- Chrome hars are built from the chrome performance logs, as seen here: https://github.com/sitespeedio/browsertime/blob/7a4e5068fdbfc86bdda37e7ca33c5fbc9da3ca81/lib/core/engineDelegate/chromeDelegate.js#L61
- To exclude preScript events (and postScript from the previous run), the performance log is emptied before loading the main page, see https://github.com/sitespeedio/browsertime/blob/7a4e5068fdbfc86bdda37e7ca33c5fbc9da3ca81/lib/core/engineDelegate/chromeDelegate.js#L51
- in this case it seems the sign-in page generates a redirect, that's picked up initially in when loading the page. However the event corresponding to the original request has been filtered out in the above step.

I recognise the current design (and perflog parsing) isn't optimal for all scenarios, and in either case it should definitely not crash. Maybe we should revisit the decision of excluding pre/post steps, however that needs some additional logic in sitespeed.io and other tools). Initially I propose two things:
- If you wait for the resulting page to load in the pre script, I think it should work as expected. That way the perflog gets emptied with no pending redirect in flight.
- Secondly we should be more robust in the parsing. Parsing the redirecting request from the redirected response will produce incorrect results. Better to (as we do in other cases where we can't find an expected request) just log a warning, skip the response and continue. I think this should work, I'll submit a PR so we can try it out.

**Comment by tobli** [comment]
Should be fixed in e7b85f0.
